det3pa/datasets/manager.py

```diff
@@ -0,0 +1,604 @@
+"""
+The manager.py module manages datasets for machine learning workflows, particularly for ``Detectron`` and ``Med3pa`` methods. 
+It includes the ``DatasetsManager`` class for handling training, validation, reference, and testing datasets, and the ``MaskedDataset`` class for different data operations.
+"""
+
+import numpy as np
+import pandas as pd
+from torch.utils.data import Dataset
+from .loading_context import DataLoadingContext
+
+class MaskedDataset(Dataset):
+    """
+    A dataset wrapper for PyTorch that supports masking and sampling of data points.
+    """
+    
+    def __init__(self, features: np.ndarray, true_labels: np.ndarray, column_labels: list = None):
+        """
+        Initializes the MaskedDataset.
+
+        Args:
+            features (np.ndarray): The feature vectors of the dataset.
+            true_labels (np.ndarray): The true labels of the dataset.
+            column_labels (list, optional): The column labels for the feature vectors. Defaults to None.
+        """
+        self.features = features
+        self.true_labels = true_labels
+        self.indices = np.arange(len(self.features))
+        self.original_indices = self.indices.copy()
+        self.sample_counts = np.zeros(len(features), dtype=int)
+        self.pseudo_probabilities = None
+        self.pseudo_labels = None 
+        self.confidence_scores = None
+        self.column_labels = column_labels if column_labels is not None else [f'feature_{i}' for i in range(features.shape[1])]
+
+    def __getitem__(self, index: int) -> tuple:
+        """
+        Retrieves the data point and its label(s) at the given index.
+
+        Args:
+            index (int): The index of the data point.
+
+        Returns:
+            tuple: A tuple containing the feature vector, pseudo label, and true label.
+        """
+        index = self.indices[index]
+        x = self.features[index]
+        y = self.true_labels[index]
+        y_hat = self.pseudo_labels[index] if self.pseudo_labels is not None else None
+        return x, y_hat, y
+
+    def __len__(self) -> int:
+        """
+        Gets the number of data points in the dataset.
+
+        Returns:
+            int: The number of data points.
+        """
+        return len(self.indices)
+    
+    def refine(self, mask: np.ndarray) -> int:
+        """
+        Refines the dataset by applying a mask to select specific data points.
+
+        Args:
+            mask (np.ndarray): A boolean array indicating which data points to keep.
+        
+        Returns:
+            int: The number of data points remaining after applying the mask.
+        
+        Raises:
+            ValueError: If the length of the mask doesn't match the number of data points.
+        """
+        if len(mask) != len(self.features):
+            raise ValueError("Mask length must match the number of data points.")
```

Comment:
One liner:
`Assert len(mask) == len(self.features), "Mask length must match the number of data points."`

Reply:
Pour le déploiement, il est mieux d'utiliser raise